docs: DOC-922,DH-18080: Document column expression validation. (#7238)
deephaven/deephaven-core: docs/groovy/conceptual/formula-validation-configuration.md

---
title: Formula validation configuration
---

When you work with Deephaven tables, you often create custom columns or filter data using expressions like:

- `myColumn = x + y * 2`
- `myTable.where("Double.isFinite(Price)")`

These expressions get sent to the Deephaven server over gRPC and compiled into executable code. For security reasons, the server needs to validate that these expressions only call approved methods and functions. Script session code execution is not validated as the Deephaven engine does not parse script code. The console service can be disabled via [configuration](../how-to-guides/configuration/console-service.md#configuration).

This guide explains how to configure formula validation using two main approaches:

- Annotations (for your own code)
- Pointcut expressions (for external libraries)

## When does validation occur?

> Formula validation happens when you:

Author: ```suggestion
Formula validation happens when a client request:
```

> - Create calculated columns using `select` or `update`.

Author: ```suggestion
- Create calculated columns using `select`, `update`, `view`, `updateView`, or `lazyUpdate`.
```

- **Group logically related methods** together rather than creating separate annotations for each method
- **Avoid creating too many small annotation categories**, as this increases configuration complexity
- **Consider the security implications** - only annotate methods that are safe for user formulas
- **Group related methods under the same annotation value** to reduce configuration

### Configuration

After adding annotations to your code, you must tell Deephaven to recognize them by setting configuration properties.

**Default configuration:**

Deephaven comes with pre-configured annotations for built-in functionality:

```properties
ColumnExpressionValidator.annotationSets.default=base,vector,function_library
```

**Adding your own annotations:**

- The property name format is: `ColumnExpressionValidator.annotationSets.<your_name>`.
- List all your annotation values separated by commas.

```properties
# Enable the annotation categories you defined
ColumnExpressionValidator.annotationSets.myapp=static_math_methods,string_helpers,safe_math,text_processing
```

## Method 2: Pointcut expressions (for external libraries)

> Use this approach when you want to allow methods from libraries you don't control (like Java's built-in classes, third-party libraries, etc.).

Author: You need to restore the link to openrewrite which has the comprehensive documentation on their format.  Telling people what the standard format is and the actual documentation is important for them to configure things correctly.

ColumnExpressionValidator.allowedMethods.integers=java.lang.Integer valueOf(int);java.lang.Integer parseInt(java.lang.String)

# Allow all methods on multiple number classes
ColumnExpressionValidator.allowedMethods.numbers=java.lang.Integer *(..);java.lang.Double *(..);java.math.BigDecimal *(..)
```

### Default Deephaven pointcut configuration

Deephaven comes pre-configured with safe methods from common Java classes:

```properties
# All methods on primitive wrapper classes (Integer, Double, etc.)
ColumnExpressionValidator.allowedMethods.primitives=java.lang.Character *(..);java.lang.Byte *(..);java.lang.Short *(..);java.lang.Integer *(..);java.lang.Long *(..);java.lang.Float *(..);java.lang.Double *(..);java.lang.Boolean *(..)

# String methods (safe because strings are immutable)
ColumnExpressionValidator.allowedMethods.basic=java.lang.String *(..)

# Number and math classes
ColumnExpressionValidator.allowedMethods.numbers=java.math.BigInteger *(..);java.math.BigDecimal *(..);java.lang.Number *(..)

# Date and time classes
ColumnExpressionValidator.allowedMethods.time=java.time.Instant *(..);java.time.LocalTime *(..);java.time.LocalDate *(..);java.time.ZonedDateTime *(..)

> # Common Object methods that are always safe

Author: # Common Object methods that are permitted

## When does validation occur?

> Formula validation happens when client API requests are sent via gRPC that:

Reviewer: ```suggestion
Formula validation happens when requests are sent over gRPC, such as those made by any of Deephaven's client APIs:
```
